File: matlab/autd3/SOEM.m
%{
%File: SOEM.m
%Project: autd3-matlab
%Created Date: 07/06/2022
%Author: Shun Suzuki
%-----
%Last Modified: 18/04/2023
%Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
%-----
%Copyright (c) 2022 Shun Suzuki. All rights reserved.
%
%}

classdef SOEM < handle

    properties
        ptr
        soem_
    end

    methods

        function obj = SOEM()
            obj.ptr = libpointer('voidPtr', 0);
            obj.soem_ = libpointer('voidPtr', 0);
            pp = libpointer('voidPtrPtr', obj.soem_);
            calllib('autd3capi_link_soem', 'AUTDLinkSOEM', pp);
        end

        function res = ifname(obj, name)
            calllib('autd3capi_link_soem', 'AUTDLinkSOEMIfname', obj.soem_, name);
            res = obj;
        end

        function res = buf_size(obj, size)
            calllib('autd3capi_link_soem', 'AUTDLinkSOEMBufSize', obj.soem_, size);
            res = obj;
        end

        function res = send_cycle(obj, cycle)
            calllib('autd3capi_link_soem', 'AUTDLinkSOEMSendCycle', obj.soem_, cycle);
            res = obj;
        end

        function res = sync0_cycle(obj, cycle)
            calllib('autd3capi_link_soem', 'AUTDLinkSOEMSync0Cycle', obj.soem_, cycle);
            res = obj;
        end

        function res = sync_mode(obj, mode)
            is_freerun = obj.sync_mode_ == SyncMode.FreeRun;
            calllib('autd3capi_link_soem', 'AUTDLinkSOEMFreerun', obj.soem_, is_freerun);
            res = obj;
        end

        function res = timer_strategy(obj, strategy)
            calllib('autd3capi_link_soem', 'AUTDLinkSOEMTimerStrategy', obj.soem_, uint8(strategy));
            res = obj;
        end

        function res = state_check_interval(obj, interval)
            calllib('autd3capi_link_soem', 'AUTDLinkSOEMStateCheckInterval', obj.soem_, interval);
            res = obj;
        end

        function res = timeout(obj, t)
            calllib('autd3capi_link_soem', 'AUTDLinkSOEMTimeout', obj.soem_, t);
            res = obj;
        end
        
        function res = build(obj)
            pp = libpointer('voidPtrPtr', obj.ptr);
            calllib('autd3capi_link_soem', 'AUTDLinkSOEMBuild', pp, obj.soem_);
            calllib('autd3capi_link_soem', 'AUTDLinkSOEMDelete', obj.soem_);
            res = obj;
        end

    end

    methods (Static)

        function adapters = enumerate_adapters()
            p = libpointer('voidPtr', 0);
            pp = libpointer('voidPtrPtr', p);
            n = calllib('autd3capi_link_soem', 'AUTDGetAdapterPointer', pp);
            adapters = strings(n, 2);

            for i = 1:n
                desc_p = libpointer('int8Ptr', zeros(128, 1, 'int8'));
                name_p = libpointer('int8Ptr', zeros(128, 1, 'int8'));
                calllib('autd3capi_link_soem', 'AUTDGetAdapter', p, i - 1, desc_p, name_p);
                desc = erase(convertCharsToStrings(char(desc_p.value)), char(0));
                name = erase(convertCharsToStrings(char(name_p.value)), char(0));
                adapters(i, :) = [desc, name];
            end

            calllib('autd3capi_link_soem', 'AUTDFreeAdapterPointer', p);
        end

    end

end

File: nim/src/autd3/soem.nim
# File: autd3.nim
# Project: src
# Created Date: 11/06/2022
# Author: Shun Suzuki
# -----
# Last Modified: 18/04/2023
# Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
# -----
# Copyright (c) 2022 Shun Suzuki. All rights reserved.
#


import strutils

import native_methods/autd3capi_link_soem
import link

type Callback* = proc(a: cstring)
type LogOut* = proc(a: cstring)
type LogFlush* = proc()


type SyncMode* {.pure.} = enum
    DC = 0
    FreeRun = 1

 
type TimerStrategy* {.pure.} = enum
    Sleep = 0
    BusyWait = 1
    NativeTimer = 2


type SOEM* = object of RootObj
  soem: pointer
  # ifname: cstring
  # bufSize: uint64
  # sync0Cycle: uint16
  # sendCycle: uint16
  # timerStrategy: TimerStrategy
  # onLost: Callback
  # syncMode: SyncMode
  # check_interval: uint64
  # debug_level: int32
  # debug_log_out: LogOut
  # debug_log_flush: LogFlush

type Adapter* = object
  name*: string
  desc*: string

func initAdapter(name: string, desc: string): Adapter =
  result.name = name
  result.desc = desc

proc enumerateAdapters*(_: typedesc[SOEM]): seq[Adapter] =
  var p = pointer(nil)
  let n = AUTDGetAdapterPointer(p.addr)
  var adapters: seq[Adapter] = @[]
  for i in 0..<n:
    var desc = cast[cstring]('\0'.repeat(128))
    var name = cast[cstring]('\0'.repeat(128))
    AUTDGetAdapter(p, i, desc, name)
    adapters.add(initAdapter($name, $desc))
  AUTDFreeAdapterPointer(p)
  adapters

func initSOEM*(): SOEM =
  AUTDLinkSOEM(result.soem.addr)

proc ifname*(cnt: var SOEM, ifname: string): var SOEM =
  AUTDLinkSOEMIfname(cnt.soem, cast[cstring](ifname))
  result = cnt

proc bufSize*(cnt: var SOEM, size: uint64): var SOEM =
  AUTDLinkSOEMBufSize(cnt.soem, size)
  result = cnt

proc sync0Cycle*(cnt: var SOEM, sync0Cycle: uint16): var SOEM =
  AUTDLinkSOEMSync0Cycle(cnt.soem, sync0Cycle)
  result = cnt

proc sendCycle*(cnt: var SOEM, sendCycle: uint16): var SOEM =
  AUTDLinkSOEMSendCycle(cnt.soem, sendCycle)
  result = cnt

proc timerStrategy*(cnt: var SOEM, strategy: TimerStrategy): var SOEM =
  AUTDLinkSOEMTimerStrategy(cnt.soem, cast[uint8](strategy))
  result = cnt

proc syncMode*(cnt: var SOEM, mode: SyncMode): var SOEM =
  AUTDLinkSOEMFreerun(cnt.soem, mode == SyncMode.FreeRun)
  result = cnt

proc onLost*(cnt: var SOEM, onLost: Callback): var SOEM =
  let p = rawProc(onLost)
  AUTDLinkSOEMOnLost(cnt.soem, p)
  result = cnt

proc stateCheckInterval*(cnt: var SOEM, intervalMs: uint64): var SOEM =
  AUTDLinkSOEMStateCheckInterval(cnt.soem, intervalMs)
  result = cnt

proc debugLevel*(cnt: var SOEM, level: int32): var SOEM =
  AUTDLinkSOEMLogLevel(cnt.soem, level)
  result = cnt
  
proc debugLogFunc*(cnt: var SOEM, logOut: LogOut, logFlush: LogFlush): var SOEM =
  let pLogOut = rawProc(logOut)
  let pLogFlush = rawProc(logFlush)
  AUTDLinkSOEMLogFunc(cnt.soem, pLogOut, pLogFlush)
  result = cnt

proc timeout*(cnt: var SOEM, timeout: uint64): var SOEM =
  AUTDLinkSOEMTimeout(cnt.soem, timeout)
  result = cnt

func build*(cnt: SOEM): Link =
  AUTDLinkSOEMBuild(result.p.addr, cnt.soem)
  AUTDLinkSOEMDelete(cnt.soem)
